GRIN→LLVM codegen: function name mismatch causes linker errors

## Context

While developing e2e tests for issue #64, a critical bug was discovered in the GRIN→LLVM code generator.

## Shortcoming

When a function is defined in GRIN with a name like `showMyBool`, the LLVM backend generates a different function name (e.g., `showMyBool.1`) during the compilation process. This causes linker errors with "undefined reference to `showMyBool`" messages.

## Deliverable

Fix the GRIN→LLVM backend to use the same function name in both the function declaration and definition.

**Example of the bug:**

```haskell
-- Source code:
showMyBool :: MyBool -> String
showMyBool MyFalse = "false"
showMyBool MyTrue = "true"
```

```grin
-- GRIN output:
main = showMyBool_2 CMyTrue_1001
-- ...
```

```
; LLVM output:
declare ptr @showMyBool(ptr)
define ptr @showMyBool.1(ptr %0) { ... }
```

**Expected behavior:** Both should use the same name.

**Impact:** This bug prevents any e2e test that defines a function with multiple pattern matches from working. It affects:
- `tests/e2e/e2e_002_bool.hs`
- `tests/e2e/e2e_005_nested_datatypes.hs`

## References

- Issue #462 (M2 Prelude work) - this is distinct from Prelude issues
- `tests/e2e/e2e_002_bool.properties` and `tests/e2e/e2e_005_nested_datatypes.properties` are currently `skip:` due to this bug

--- a/src/backend/grin_to_llvm.zig
+++ b/src/backend/grin_to_llvm.zig
@@ -599,7 +599,10 @@ pub const GrinTranslator = struct {
 
         const fn_type = llvm.functionType(ret_type, param_types[0..def.params.len], false);
         const fn_name_z = self.formatName(def.name);
-        const func = llvm.addFunction(self.module, fn_name_z, fn_type);
+        // Check if function was already forward-declared; if so, reuse it.
+        // This prevents LLVM from adding .1 suffix due to name collision.
+        const func = c.LLVMGetNamedFunction(self.module, fn_name_z) orelse
+            llvm.addFunction(self.module, fn_name_z, fn_type);
         self.current_func = func;
         const entry_bb = llvm.appendBasicBlock(func, "entry");
         llvm.positionBuilderAtEnd(self.builder, entry_bb);

--- a/tests/e2e/e2e_002_bool.properties
+++ b/tests/e2e/e2e_002_bool.properties
@@ -1,1 +1,4 @@
-skip: GRIN→LLVM function name mismatch causes linker errors - tracked in #463
+# Test 002: Inductive data types with multiple constructors
+# Uses standard Bool type (True/False) to test that user-defined types
+# can shadow Prelude built-in constructors without duplicate errors
+exit_code: 0

--- a/tests/e2e/e2e_005_nested_datatypes.properties
+++ b/tests/e2e/e2e_005_nested_datatypes.properties
@@ -1,1 +1,3 @@
-skip: GRIN→LLVM function name mismatch causes linker errors - tracked in #463
+# Test 005: Nested data types (Maybe and List)
+# Tests nested data constructor handling
+exit_code: 0

